File: 137. Single Number II/singleNumber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const singleNumber = (nums) => {
  let ones = 0;
  let twos = 0;

  for (const num of nums) {
    ones = (ones ^ num) & ~twos;
    twos = (twos ^ num) & ~ones;
  }

  return ones;
};


const nums = [2, 2, 3, 2];
console.log('singleNumber: ', singleNumber(nums)); // 3

File: 137. Single Number II/singleNumber.py
from typing import List

class Solution:
  def singleNumber(self, nums: List[int]) -> int:
    ones = 0
    twos = 0

    for num in nums:
      ones = (ones ^ num) & ~twos
      twos = (twos ^ num) & ~ones

    return ones
  

nums = [2, 2, 3, 2]
solution = Solution()
print('singleNumber: ', solution.singleNumber(nums)) # 3
